### PR TITLE
development for project5

### PR DESCRIPTION
**Pull request by AaronMoroney**

### CODE REVIEW

File: code/front/js/index.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+
+/*
+** index page (all products)
+*/
+
+const uri = 'http://localhost:3000/api/products/';
```
Comment (fpatelm) [inline code/front/js/index.js]: best practice to declare global const variables all upper case, in this case:
`const URI = 'http://localhost:3000/api/products/';`

File: code/front/js/index.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+const singleProductLink = './productFront.html?=';
```
Comment (fpatelm) [inline code/front/js/index.js]: best practice to declare global const variables all upper case, in this case:
const SINGLE_PRIODUCT_LLINK= './productFront.html?=';

File: code/front/js/index.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+});
+    
+
+/*
+** | create product cards 
+*/
+
+function createProductCardsInfo(array) {
+    const length = array.length;
+    for (let i = 0; i<length; i++) {
```
Comment (fpatelm) [inline code/front/js/index.js]:
```
 for (let i = 0; i<array.length; i++) {
}

```

File: code/front/js/index.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    items.appendChild(pageLink); //link  to parent
+    pageLink.appendChild(article); //product description parent to link
+
+    //append product descriptors to article ( in order )
+    article.appendChild(img);
+    article.appendChild(productName);
+    article.appendChild(productDescription);
+    
+    return createProductCardView;
```
Comment (fpatelm) [inline code/front/js/index.js]: why are you returning this view that is not used in the calling method?
see line #28, ` createProductCardView(array[i]);`
you are not using the returned value.

File: code/front/js/index.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+
```
Comment (fpatelm) [inline code/front/js/index.js]: remove all this empty lines

File: code/front/js/productFront.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+** single product page
+*/
+
+//get _id of product from query parameters
+const queryString = window.location.search;
+console.log(queryString);
+let splitForId = queryString.slice(5);
+console.log(splitForId);
+
+const newUri = 'http://localhost:3000/api/products/' + splitForId;
```
Comment (fpatelm) [inline code/front/js/productFront.js]: make it:
`const PRODUCT_URI  = 'http://localhost:3000/api/products/' + PRODUCT_ID;`

File: code/front/js/productFront.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
+/*
+** single product page
+*/
+
+//get _id of product from query parameters
+const queryString = window.location.search;
+console.log(queryString);
+let splitForId = queryString.slice(5);
```
Comment (fpatelm) [inline code/front/js/productFront.js]: make it
`const PRODUCT_ID = queryString.slice(5);`

File: code/front/js/productFront.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+const newUri = 'http://localhost:3000/api/products/' + splitForId;
+let product;
+
+/*
+** fetch
+*/
+
+//fetch the new URI from get request
+fetch(newUri)
```
Comment (fpatelm) [inline code/front/js/productFront.js]: `fetch(PRODUCT_URI)`

File: code/front/js/productFront.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+*/
+
+//fetch the new URI from get request
+fetch(newUri) 
+    .then((response) => response.json())
+    .then((data) => {
+        console.log(data)
+        createIndividualProductView(data);
+        populateDropdown(data);
+        product = data;
```
Comment (fpatelm) [inline code/front/js/productFront.js]: ` product = data; `should be the first thing!

```
console.log(data)
product = data;
createIndividualProductView(data);
populateDropdown(data);
```

File: code/front/js/productFront.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    //populate 
+    individualProductName.innerText = data.name;
+    individualProductPrice.innerText = data.price;
+    individualProductDescription.innerText = data.description;
+    individualProductImg.src = data.imageUrl;
+
+    //append
+    imgParent.appendChild(individualProductImg);
+
+    return createIndividualProductView;
```
Comment (fpatelm) [inline code/front/js/productFront.js]: again,, you are returning a value and not using it ....., see line #23, if you do use it at the calling method, don't return it.

read about java functions, your evaluator will ask this, and you can fail the presentation if you don't know how to explain when to use a return value or not.

File: code/front/js/productFront.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    alert('please select a value');
+    return; 
+  }
+  console.log(interestedProduct); //array
+  //call function
+  addToShoppingCart(interestedProduct); 
+});
+
+/*
+** | add to shopping cart
```
Comment (fpatelm) [inline code/front/js/productFront.js]: add more meaningful comments in all the functions:
what does the function receive?
what does the function do?
what does the function return?

File: code/front/js/productFront.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    foundProduct[0].quantity = interestedProduct.quantity + foundProduct[0].quantity;
+    console.log(foundProduct[0]);
+    //cart.push(interestedProduct);
+    console.log(cart);
+    //re-store data in local storage
+    localStorage.setItem('scart', JSON.stringify (cart));
+  }
+}
+
+
```
Comment (fpatelm) [inline code/front/js/productFront.js]: r3emove extra empty lines

**Review by fpatelm — CHANGES_REQUESTED** [review]

File: code/front/js/index.js (hunk 4 above)
Comment (fpatelm) [reply code/front/js/index.js]: @AaronMoroneyDesignCo  this was not resolved!

File: code/front/js/index.js (hunk 5 above)
Comment (fpatelm) [reply code/front/js/index.js]: you did not remove empty lines fo code

File: code/front/js/index.js (hunk 3 above)
Comment (fpatelm) [reply code/front/js/index.js]: you did not update it

File: code/front/js/productFront.js (hunk 8 above)
Comment (AaronMoroney) [reply code/front/js/productFront.js]: This was resolved but didn't successfully appear in the git push to dev branch, Perhaps I should hit command+s first on the files

File: code/front/js/index.js (hunk 5 above)
Comment (AaronMoroney) [reply code/front/js/index.js]: Also had been previously resolve on my local codebase

File: code/front/js/index.js (hunk 3 above)
Comment (AaronMoroney) [reply code/front/js/index.js]: now updated

File: code/front/js/index.js (hunk 4 above)
Comment (AaronMoroney) [reply code/front/js/index.js]: Also had been resolved - is it standard to hit command + S before a push to branch?

File: code/front/js/cart.js [hunk 13]
```diff
@@ -113,7 +113,7 @@ function populateCartInfo(cart) {
 function updateCartTotal () {
     //totals
     //only need 1, class returns an array of hmtl so, [0]
-    let cartPrice = document.getElementsByClassName('cart__price')[0];
+    document.getElementsByClassName('cart__price')[0];
```
Comment (fpatelm) [inline code/front/js/cart.js:116]: this code is not doing anything

File: code/front/js/cart.js [hunk 14]
```diff
@@ -205,8 +228,13 @@ function updateQuantityCartItem () {
             inputListener.getAttribute('value');
             //cart loop
             for (let i = 0; i < cart.length; i++) {
-                itemQuantity.value = event.target.value;
-                cart[i].quantity = itemQuantity[i].value;
+                if (cart.length === 1) {
```
Comment (fpatelm) [inline code/front/js/cart.js:231]: here you have to use the same approach as for deletecart item

imagine you update the "fat milk 2l" item on the cart and on the cart you have as well "fat milk 1l". You need to know which item you want to update the quantity. You need to get the respective item by id and color from the cart and update its quantity.
```
let updatedItem = event.target;
 let color = updatedItem.closest('.cart__item').getAttribute('data-color');
let productId = updatedItem.closest('.cart__item').getAttribute('data-id');
```

**Review by fpatelm — CHANGES_REQUESTED** [review]

File: code/front/js/cart.js (hunk 14 above)
Comment (AaronMoroney) [reply code/front/js/cart.js:231]: Hey Faizal, Just to clarify - if I add one product ID with two different colour options -  (a) white (b) black - the product Total in cart = 2. The products are stored in the cart array as two separate objects(products) like so - [  {productID: "1" colour: "white"}, { productID: "2" color: "black" } ] 

Is it necessary to target the data-ID attribute, and the data-color attribute ? I can successfully update the products in the cart without doing so as it stands - and all tests pass.

Will this refactoring suggestion will allow me **not** to have to wrap the logic for updateQuantityCartItem in if else? I don't understand.

Comment (fpatelm) [reply code/front/js/cart.js:231]: will mark this as unnecessary as we discussed in the workplace

File: code/front/js/cart.js (hunk 13 above)
Comment (fpatelm) [reply code/front/js/cart.js:116]: you can just remove this line of code

**Review by fpatelm — APPROVED** [review]

**Pull request merged**